compiler-plugin/src/main/java/io/ballerina/stdlib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,396 @@
+            case RECORD, ARRAY, TUPLE, MAP, UNION, INTERSECTION -> {
+                return true;
+            }
+            default -> {
+                return false;
+            }
+        }
+    }
+
+    private boolean isSupportedExpectedType(TypeSymbol typeSymbol) {
+        TypeDescKind kind = typeSymbol.typeKind();
+        if (kind == TypeDescKind.TYPE_REFERENCE) {
+            kind = ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor().typeKind();
+        }
+
+        switch (kind) {
+            case ARRAY, UNION, INTERSECTION -> {
+                return true;
+            }
+            default -> {
+                return false;
+            }
+        }
+    }
+
+    public static TypeSymbol getRawType(TypeSymbol typeDescriptor) {
+        if (typeDescriptor.typeKind() == TypeDescKind.INTERSECTION) {
+            return getRawType(((IntersectionTypeSymbol) typeDescriptor).effectiveTypeDescriptor());
+        }
+        if (typeDescriptor.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+            TypeReferenceTypeSymbol typeRef = (TypeReferenceTypeSymbol) typeDescriptor;
+            if (typeRef.typeDescriptor().typeKind() == TypeDescKind.INTERSECTION) {
+                return getRawType(((IntersectionTypeSymbol) typeRef.typeDescriptor()).effectiveTypeDescriptor());
+            }
+            TypeSymbol rawType = typeRef.typeDescriptor();
+            if (rawType.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+                return getRawType(rawType);
+            }
+            return rawType;
+        }
+        return typeDescriptor;
+    }
+
+    private void reportDiagnosticInfo(SyntaxNodeAnalysisContext ctx, Optional<Location> location,
+                                      CsvDataDiagnosticCodes diagnosticsCodes) {
+        Location pos = location.orElseGet(() -> currentLocation);
+        DiagnosticInfo diagnosticInfo = new DiagnosticInfo(diagnosticsCodes.getCode(),
+                diagnosticsCodes.getMessage(), diagnosticsCodes.getSeverity());
+        if (allDiagnosticInfo.containsKey(pos) && allDiagnosticInfo.get(pos).equals(diagnosticInfo)) {
+            return;
+        }
+        allDiagnosticInfo.put(pos, diagnosticInfo);
+        ctx.reportDiagnostic(DiagnosticFactory.createDiagnostic(diagnosticInfo, pos));
+    }
+
+    private void processModuleVariableDeclarationNode(ModuleVariableDeclarationNode moduleVariableDeclarationNode,
+                                                      SyntaxNodeAnalysisContext ctx) {
+        Optional<ExpressionNode> initializer = moduleVariableDeclarationNode.initializer();
+        if (initializer.isEmpty() || !isParseFunctionOfStringSource(initializer.get())) {
+            return;
+        }
+
+        Optional<Symbol> symbol = semanticModel.symbol(moduleVariableDeclarationNode.typedBindingPattern());
+        if (symbol.isEmpty()) {
+            return;
+        }
+        validateExpectedType(((VariableSymbol) symbol.get()).typeDescriptor(), ctx);
```

Comment:
Prefer monadic style over explicit checks when it comes to optionals. Check other places you have used optionals as well
```suggestion
        symbol.map(s -> (VariableSymbol) s).map(s -> typeDescriptor()).ifPresent(s -> validateExpectedType(s, ctx);
```